`BlockHash`: a new distinct type solely for block hashes

dahlia opened the pull request:
This closes <https://github.com/planetarium/libplanet/issues/1192>.

The most changes are simple renaming from `HashDigest<SHA256>` to `BlockHash`, but unlike `HashDigest<SHA256>`, as `BlockHash` is variable-length (hence `BlockHash.BytesLength` property), some code needed to explicitly deal with block hashes' lengths.

See also the changelog.

moreal reviewed (commented):
I will review again to check changelog was written well for covering all changes.. 🤔

dahlia commented on Libplanet/HashDigest.cs:
```diff
@@ -69,19 +69,15 @@ public HashDigest(byte[] hashDigest)
+        /// <exception cref="ArgumentOutOfRangeException">Thrown when the given
+        /// <paramref name="hashDigest"/>'s <see cref="ImmutableArray{T}.Length"/> is not
+        /// the same to the <see cref="Size"/> the hash algorithm
+        /// (i.e., <typeparamref name="T"/> requires.</exception>
```
Thanks! Fixed: 591379046335ae121ec378f8f3862f2f8a7e1514.

dahlia commented on Libplanet/HashDigest.cs:
```diff
@@ -34,17 +35,21 @@ namespace Libplanet
-            Size = thunk.HashSize / 8;
-
+            Type type = typeof(T);
+            MethodInfo method = type.GetMethod("Create", new Type[0])!;
```
Thanks! Fixed: fb2410b0f8125e573bb5e1f81d63bddbd706bb2a.

dahlia commented:
@planetarium/libplanet Rebased on the latest main.  Please take a look again!

longfin commented on Libplanet/Store/BaseStore.cs:
```diff
@@ -23,23 +23,21 @@ public abstract class BaseStore : IStore, IDisposable
+        public abstract long AppendIndex(Guid chainId, BlockHash hash);
 
-        /// <inheritdoc />
+        /// <inheritdoc cref="IStore.ForkBlockIndexes(Guid, Guid, BlockHash)"/>
```
is `cref` in `inheritdoc` needed?

dahlia replied:
It's not mandatory in theory, but since I've observed [DocFX] did not render `<inheritdoc>`s without `cref`, I tend to give them explicit `cref`.

[DocFX]: https://dotnet.github.io/docfx/

longfin commented on Libplanet/Store/Trie/MerkleTrie.cs:
```diff
@@ -27,8 +27,8 @@ public class MerkleTrie : ITrie
             _codec = new Codec();
-            EmptyRootHash = Hashcash.Hash(
-                _codec.Encode(default(Null)!));
+            var bxNull = _codec.Encode(default(Null)!);
```
```suggestion
            byte[] bxNull = _codec.Encode(default(Null)!);
```

longfin commented on Libplanet/Tx/Transaction.cs:
```diff
@@ -446,7 +451,7 @@ public static Transaction<T> Deserialize(byte[] bytes, bool validate = true)
                     actionsArray
                 ).EvaluateActions(
-                    default(HashDigest<SHA256>),
+                    emptyBlockHash,
```
Can't we use just `default(BlockHash)` here?

dahlia replied:
No, since `BlockHash` is unsized, `default(BlockHash)` resembles `new byte[0]` in memory layout, whereas `default(HashDigest<SHA256>)` resembles `new byte[32]`.

longfin replied:
👌

dahlia commented on CHANGES.md:
```diff
@@ -8,6 +8,116 @@ To be released.
+        (was `HashDigest<SHA256>?`).
+     -  The type of `IStore.AppendIndex()` method's `hash` parameter became
+        `BlockHash` (was `HashDigest<SHA256>`).
+     -  Replaced `IStore.ForkBlockIndexes()` method's
+        `HashDigest<SHA256> branchPoint` parameter with `BlockHash branchpoint`.
```
This does not change only the parameter type, also the parameter name (`branchPoint` → `branchpoint`).

dahlia commented on CHANGES.md:
```diff
@@ -78,6 +188,8 @@ To be released.
  -  Added `Block<T>.Header` property.  [[#1070], [#1102]]
  -  Added `BlockHeader.ProtocolVersion` property.  [[#1142], [#1147]]
+ -  Added `BlockHash` struct.  [[#1192], [#1197]]
+ -  Added `HashDigest<T>.DeriveFrom()` method.  [[#1197]]
```
This was intended.

dahlia commented:
@planetarium/libplanet Rebased on the up-to-date main branch.  Please take a look again.

longfin commented:
It seems that we need to resolve conflicts... 😢

dahlia commented:
@planetarium/libplanet Resolved some conflicts, and rebased on the up-to-date main branch.  Please take a look again.

riemannulus commented:
Is this CI broken belong to turn server?

dahlia commented:
> Is this CI broken belong to TURN server?

Yes, the build failed due to the TURN server it depends on.  #1243 addressed the problem, and it's now fixed as I rebased.

dahlia commented:
@planetarium/libplanet Rebased, and the build is green.  Please take a look again.
